### PR TITLE
Support retries on conflicting image stream mapping create attempts

### LINKED ISSUE #5421: TestTriggers is flaky
```
--- FAIL: TestTriggers_imageChange-2 (6.15s)
    deploy_trigger_test.go:164: unexpected error: imageStream "test-image-stream" cannot be updated: the object has been modified; please apply your changes to the latest version and try again
FAIL
```

@ironcladlou

### PR DESCRIPTION
**Pull request by ironcladlou**
Add retry support to mapping creation to allow repeated update attempts
when the conflicting resource doesn't contain any tag diffs.

Fixes #5421

### CODE REVIEW

**Comment by ironcladlou** [comment]
@deads2k @smarterclayton @ncdc PTAL.

**Comment by ironcladlou** [comment]
[test]

**Comment by deads2k** [comment]
Little bit of plumbing for that test, eh?

lgtm

**Comment by ironcladlou** [comment]
>Little bit of plumbing for that test, eh?

:weary:

File: pkg/image/registry/imagestreammapping/rest.go [hunk 1]
```diff
@@ -90,18 +102,52 @@ func (s *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
...
+			// nothing actually changed
+			result = &unversioned.Status{Status: unversioned.StatusSuccess}
+			return nil
+		}
+		api.UpdateTrackingTags(stream, tag, next)
 
-	api.UpdateTrackingTags(stream, tag, next)
+		_, err := s.imageStreamRegistry.UpdateImageStreamStatus(ctx, stream)
+		if err != nil {
+			if !errors.IsConflict(err) {
```
Comment (smarterclayton) [inline pkg/image/registry/imagestreammapping/rest.go]: If you have this then there is no point to having RetryOnConflic.t... which is supposed t handle conflict errors.

**Comment by smarterclayton** [comment]
Hold on LGTM

**Comment by smarterclayton** [comment]
Not using RetryOnConflict correctly.

File: pkg/image/registry/imagestreammapping/rest.go [hunk 2]
```diff
@@ -90,18 +102,52 @@ func (s *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
...
+				return err
+			}
+			// If the update conflicts, get the latest stream and check for tag
```
Comment (smarterclayton) [inline pkg/image/registry/imagestreammapping/rest.go]: All of this code should be at the beginning of the loop - guarded by "don't do this on the first loop".  Before the call to RetryOnConflict, set the last stream tag.  At the beginning of the loop, check whether stream == nil.  Then check whether old tag == new tag.  If they don't, then exit.  Do the mutation.  If you get *any* error, clear stream (so the next iteration of the loop reads it again).

File: pkg/image/registry/imagestreammapping/rest.go (hunk 1 above)
Comment (ironcladlou) [inline pkg/image/registry/imagestreammapping/rest.go]: Not sure I understand. If the error is *not* a conflict error, return early and pass that on to the client (i.e. no retry). If it *is* a conflict, do a deeper inspection to decide whether to allow the helper to deal with the conflict (i.e. retry).

File: pkg/image/registry/imagestreammapping/rest.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/image/registry/imagestreammapping/rest.go]: If you want to terminate early, return a known error type that you convert to conflict on the outside.

Comment (deads2k) [inline pkg/image/registry/imagestreammapping/rest.go]: > If you get any error, clear stream (so the next iteration of the loop reads it again).

This code relies on not resetting the stream unless it finds there isn't a real conflict.  That logic ensures that it never re-attempts this work without an imagestream with a resourceversion that is known to be consistent with the first one we grabbed.

**Comment by deads2k** [comment]
> Not using RetryOnConflict correctly.

I find the current code easier to inspect that the changes needed for `RetryOnConflict`.  Maybe just not use that construct?

File: pkg/image/registry/imagestreammapping/rest.go (hunk 2 above)
Comment (ironcladlou) [inline pkg/image/registry/imagestreammapping/rest.go]: Using a custom error type makes sense to clean up that little mess, but I'm still trying to understand what's being proposed for the rest of the refactor. I don't really see how what I have is wrong.

**Comment by ironcladlou** [comment]
I don't think the conflict-based aspect of the retry helper is buying us anything in this case since we have to micromanage the conflict error type ourselves (while the helper wants to assume that all conflicts mean "retry").

What would be more appropriate IMO is a generic retry/backoff helper that lets me return a bool indicating whether or not to retry (sort of like `wait.Poll`).

**Comment by smarterclayton** [comment]
If you don't like RetryOnConflict, this isn't the time to change it, but if
it's not solving the problem then it needs to be changed.

This code, as written, isn't really using RetryOnConflict the way its
intended to be used.

**Comment by smarterclayton** [comment]
You can use wait.ExponentialBackoff and pass the same arguments, but you'll
get your early exit.

**Comment by ironcladlou** [comment]
Refactored to use `wait.ExponentialBackoff`; looks much cleaner in this case IMO.

File: pkg/image/registry/imagestreammapping/rest.go [hunk 3]
```diff
@@ -90,17 +99,38 @@ func (s *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
...
+			// nothing actually changed
+			return true, nil
+		}
+		api.UpdateTrackingTags(stream, tag, next)
+		_, err := s.imageStreamRegistry.UpdateImageStreamStatus(ctx, stream)
+		if err == nil {
+			return true, nil
+		}
+		if !errors.IsConflict(err) {
+			return true, err
```
Comment (liggitt) [inline pkg/image/registry/imagestreammapping/rest.go]: `false, err`, right?

Comment (ironcladlou) [inline pkg/image/registry/imagestreammapping/rest.go]: Not sure that it matters- ExponentialBackoff will return due to err != nil. Technically we're also "done" at this point...

File: pkg/image/registry/imagestreammapping/rest.go [hunk 4]
```diff
@@ -90,17 +99,38 @@ func (s *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
...
+			return true, findLatestErr
+		}
+		newerEvent := api.LatestTaggedImage(latestStream, tag)
+		if lastEvent == nil || kapi.Semantic.DeepEqual(lastEvent, newerEvent) {
+			// The tag hasn't changed, so try again with the updated stream.
+			stream = latestStream
+			return false, nil
+		}
+		// The tag changed, so return the conflict error back to the client.
+		return true, err
```
Comment (smarterclayton) [inline pkg/image/registry/imagestreammapping/rest.go]: If you return err, you shouldn't return true.  Err overrides true.

**Comment by soltysh** [comment]
I was recently faced with similar problem and tried to stay away from modifying the storage layer, but rather was inclined to that in the client code which should support situations as such. My reasoning was the storage should be as simple as possible, without any retry logic. Although code looks good, I'm not too happy with this change. I'm afraid it might hide some bigger problems...

**Comment by liggitt** [comment]
retry for conflict on create is weird... we shouldn't make clients do that

**Comment by soltysh** [comment]
The root cause of this issue is we're trying to create ISM and update IS status at one shot, which ends up with those failures. Maybe in that case we should rather fix in `UpdateImageStreamStatus` and not here?

**Comment by soltysh** [comment]
Whereas most, if not all (except for this one and I think I know other one), of our storage are dealing with single object. Wherever more than one come into play, problems occur.

**Comment by smarterclayton** [comment]
This is an idempotent API operation (add mapping).  So retry at the server is appropriate.

**Comment by soltysh** [comment]
OK, once this gets in, I'll revert one of the recent changes I did in https://github.com/openshift/origin/blob/master/pkg/dockerregistry/server/repositorymiddleware.go#L207-L212 as it'll be unnecessary.

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**